core/shared/src/test/scala/fs2/StreamSuite.scala

```diff
@@ -1024,6 +1024,20 @@ class StreamSuite extends Fs2Suite {
     assert(compileErrors("Stream.eval(IO(1)).through(p)").nonEmpty)
   }
 
+  test("Unintended unchunking when using StreamLowPriority#monadInstance") {
+    def generateTuple(pipe: Pipe[fs2.Pure, Int, (Int, Int)]) =
+      (Stream(1, 2, 3) ++ Stream(4, 5, 6)).through(pipe).chunks.toList
+    assertEquals(generateTuple(_.map(e => e -> (e + 1))), generateTuple(_.fproduct(_ + 1)))
+  }
+
+  test("Unintended unchunking when using Stream#monadErrorInstance") {
+    def generateTuple(pipe: Pipe[IO, Int, (Int, Int)]) =
+      (Stream(1, 2, 3) ++ Stream(4, 5, 6)).covary[IO].through(pipe).chunks.compile.toList
+    val lhs = generateTuple(_.map(e => e -> (e + 1)))
+    val rhs = generateTuple(_.fproduct(_ + 1))
+    (lhs, rhs).mapN(assertEquals(_, _))
+  }
+
```

Comment:
These tests seem to be needlessly complicated, with the use of different literal values, the tuple, the increment, etc. 

In a "test-driven-development" fashion, the process is to 1. write the smallest possible tests that fails 2. change code, 3. check that test passes. It seems to me that would be achieved, for the problem, with a test core that did the following: 

```Scala
def countChunks(str: Stream[Nothing, _]): Int = 
  str.chunks.toList.length

val source = Stream(0) ++ Stream(0, 0)
val mapped = source.map(_ => 1)
assertEquals(countChunks(mapped), 2)
```

Reply:
Thanks for your review!
It's helpful, and I will make a change for this:)

Reply:
Hi, sorry if I made some mistakes.
I run a local test and found that only calling the 'map' method like `source.map(_ => 1)` can't call the method I changed. :)
I changed my code to below to make it simpler, is it OK?
```scala
  test("Unintended unchunking when using StreamLowPriority#monadInstance") {
    def countChunks(): Int =
      (Stream(0) ++ Stream(0, 0)).fproduct(_ + 1).chunks.toList.length
    assertEquals(countChunks(), 2)
  }

  test("Unintended unchunking when using Stream#monadErrorInstance") {
    def countChunks() =
      (Stream(0) ++ Stream(0, 0)).covary[IO].fproduct(_ + 1).chunks.compile.toList.map(_.length)
    countChunks().assertEquals(2)
  }
```

Reply:
Actually, we can do even better.
```
Stream.monadInstance.map(source)(_ => 1)
// etc.
```

Reply:
Sorry, but I encountered a problem. :(
`Stream.monadInstance.map(source)` works well, but for `Stream.monadErrorInstance.map(str)`, my IDE shows 'No implicits found for parameter ev: ApplicativeError[F_ , Throwable]', what's that? 😂

Reply:
Thanks for trying that! If you make the change I suggested in https://github.com/typelevel/fs2/pull/3189#discussion_r1154558125, then we only need one test.

Reply:
OK, but can it cover the 'MonadErrorInstance'? 🤔

Reply:
Yes, because my proposal is that `Monad` and `MonadError` implementations should share the same code. Then we don't need to write the same code twice, fix the same bugs twice, and add the same tests twice :)